Begin integrating password autofill with LoginStatus API

pooneh-nb opened the pull request:
#### 68291cef7f3bf33d44e16f0a6bfc2770988bc65e
<pre>
Begin integrating password autofill with LoginStatus API
<a href="https://bugs.webkit.org/show_bug.cgi?id=279084">https://bugs.webkit.org/show_bug.cgi?id=279084</a>
<a href="https://rdar.apple.com/135221840">rdar://135221840</a>

Reviewed by NOBODY (OOPS!).

We integrate Password Autofill with LoginStatus API. So setting a status is now conditional to logging in getting a signal from password Autofill in 30s. We receive this signal from Safari indicating whether the user has clicked the password autofill option.

Originally written by pooneh-nb &lt;pnikkhahbahrami@ucdavis.edu&gt;

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setLastAuthentication):
* Source/WebCore/page/Page.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setLoginStatus):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _didFillPasswordForUsername:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFillPasswordForUsername):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLoginStatus):
(WebKit::WebPage::didFillPasswordForUsername):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST(ResourceLoadStatistics, LoginStatusRequiresPasswordAutoFill)):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68291cef7f3bf33d44e16f0a6bfc2770988bc65e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96743 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16357 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47263 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98788 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16653 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24797 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73728 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30946 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99746 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12539 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87546 "Found 1 new API test failure: TestWebKitAPI.ResourceLoadStatistics.LoginStatusRequiresPasswordAutoFill (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54063 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12298 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5361 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46591 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82396 "Found 1 new API test failure: TestWebKitAPI.ResourceLoadStatistics.LoginStatusRequiresPasswordAutoFill (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5443 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103839 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23811 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/17368 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82776 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24062 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83616 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82163 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26821 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4396 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/17289 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23773 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28928 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23432 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26912 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25173 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1191,18 +1191,29 @@ void NetworkConnectionToWebProcess::requestStorageAccess(RegistrableDomain&& sub
-        if (umanagedAuthentication.hasException())
-            return completionHandler();
-        lastAuthentication = umanagedAuthentication.releaseReturnValue();
+    std::optional<WebCore::LoginStatus> trustWorthySignal;
```
"trustworthy" is a single word, so no capital `W`

cdumez commented on Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
```diff
@@ -1344,6 +1344,11 @@ - (BOOL)providesWritingToolsContextMenu
 
+- (void)didFillPasswordForUsername:(NSString *)username
+{
+    _page->didFillPasswordForUsername(String(username));
```
I don't think you need the `String()` here

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -787,6 +787,8 @@ typedef NS_OPTIONS(NSUInteger, WKDisplayCaptureSurfaces) {
 
 @property (nonatomic, readonly) CGFloat minimumMagnification WK_API_AVAILABLE(macos(15.0));
 
+- (void)didFillPasswordForUsername:(NSString *)username;
```
This needs to start with an underscore

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2888,6 +2892,8 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 
     mutable RefPtr<Logger> m_logger;
+
+    std::optional<WebCore::LoginStatus> m_lastPasswordAutoFill;
```
`std::optional<WebCore::LoginStatus> m_lastAuthentication` already exists on `Page` - we should be able to use that and remove this.

pooneh-nb committed:
Begin integrating password autofill with LoginStatus API
https://bugs.webkit.org/show_bug.cgi?id=279084
rdar://135221840

Reviewed by NOBODY (OOPS!).

We integrate Password Autofill with LoginStatus API. So setting a status is now conditional to logging in getting a signal from password Autofill in 30s. We receive this signal from Safari indicating whether the user has clicked the password autofill option.

Originally written by pooneh-nb <pnikkhahbahrami@ucdavis.edu>

* Source/WebCore/Modules/webauthn/AuthenticatorCoordinator.cpp:
(WebCore::AuthenticatorCoordinator::discoverFromExternalSource):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setLastAuthentication):
* Source/WebCore/page/Page.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::setLoginStatus):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _didFillPasswordForUsername:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFillPasswordForUsername):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setLoginStatus):
(WebKit::WebPage::didFillPasswordForUsername):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST(ResourceLoadStatistics, LoginStatusRequiresPasswordAutoFill)):

charliewolfe commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -2174,3 +2175,31 @@ HTTPServer httpServer({
+
+    [webView evaluateJavaScript:@"navigator.setStatus('logged-out');" completionHandler:^(id, NSError *) {
+        [webView evaluateJavaScript:@"navigator.isLoggedIn()" completionHandler:^(id result, NSError *) {
+            EXPECT_FALSE([result boolValue]);
```
This test needs to wait until the completion handlers are called before finishing. See the uses of `TestWebKitAPI::Util::run(&done);` in this file for an example.

charliewolfe commented on Source/WebCore/page/Page.cpp:
```diff
@@ -5042,9 +5042,12 @@ void Page::activeNowPlayingSessionUpdateTimerFired()
+    String usernameToSet = emptyString();
+    if (!username.isEmpty())
+        usernameToSet = username;
+    auto loginStatus = LoginStatus::create(RegistrableDomain(mainFrameURL()), usernameToSet, LoginStatus::CredentialTokenType::HTTPStateToken, authType, LoginStatus::TimeToLiveAuthentication);
```
I don't think we need the variable and if statement, could just be:
```suggestion
    auto loginStatus = LoginStatus::create(RegistrableDomain(mainFrameURL()), username, LoginStatus::CredentialTokenType::HTTPStateToken, authType, LoginStatus::TimeToLiveAuthentication);
```

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -1815,7 +1817,6 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
 #endif
 
     void setTopContentInset(float);
-
```
undo this change

charliewolfe commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -290,6 +290,8 @@ for this property.
 - (void)_isJITEnabled:(void(^)(BOOL))completionHandler WK_API_AVAILABLE(macos(10.14.4), ios(12.2));
 - (void)_removeDataDetectedLinks:(dispatch_block_t)completion WK_API_AVAILABLE(macos(10.14.4), ios(12.2));
 
+- (void)_didFillPasswordForUsername:(NSString *)username;
```
This SPI needs availability clauses: `WK_API_AVAILABLE(macos(WK_MAC_TBA), ios(WK_IOS_TBA))`

charliewolfe commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -219,6 +219,7 @@ class IntRect;
 class KeyboardEvent;
 class LocalFrame;
 class LocalFrameView;
+class LoginStatus;
```
This doesn't look like it's used, should be removed.

samuelgoto commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
```diff
@@ -2201,3 +2202,39 @@ HTTPServer httpServer({
+    [webView _didFillPasswordForUsername:@"example"];
+
+    __block bool done { false };
+    [webView evaluateJavaScript:@"navigator.setStatus('logged-in');" completionHandler:^(id, NSError *) {
```
Any chance you can move these under the navigator.login type?

```suggestion
    [webView evaluateJavaScript:@"navigator.login.setStatus('logged-in');" completionHandler:^(id, NSError *) {
```

For reference, here is what we have implemented in Chrome:

https://w3c-fedid.github.io/login-status/#dom-navigatorlogin-setstatus

samuelgoto replied:
@johnwilander FYI
